Mathlib/Data/Finsupp/MonomialOrder/DegLex.lean

```diff
@@ -0,0 +1,261 @@
+instance : LinearOrder (DegLex (α →₀ ℕ)) :=
+  LinearOrder.lift'
+    (fun (f : DegLex (α →₀ ℕ)) ↦ toLex ((ofDegLex f).degree, toLex (ofDegLex f)))
+    (fun f g ↦ by simp)
+
+theorem le_iff {x y : DegLex (α →₀ ℕ)} :
+    x ≤ y ↔ (ofDegLex x).degree < (ofDegLex y).degree ∨
+      (ofDegLex x).degree = (ofDegLex y).degree ∧ toLex (ofDegLex x) ≤ toLex (ofDegLex y) := by
+  simp only [le_iff_eq_or_lt, lt_iff, EmbeddingLike.apply_eq_iff_eq]
+  by_cases h : x = y
+  · simp [h]
+  · by_cases k : (ofDegLex x).degree < (ofDegLex y).degree
+    · simp [k]
+    · simp only [h, k, false_or]
+
+noncomputable instance : OrderedCancelAddCommMonoid (DegLex (α →₀ ℕ)) where
+  le_of_add_le_add_left a b c h := by
+    rw [le_iff] at h ⊢
+    simpa only [ofDegLex_add, degree_add, add_lt_add_iff_left, add_right_inj, toLex_add,
+      add_le_add_iff_left] using h
+  add_le_add_left a b h c := by
+    rw [le_iff] at h ⊢
+    simpa [ofDegLex_add, degree_add] using h
+
+/-- The linear order on `Finsupp`s obtained by the homogeneous lexicographic ordering. -/
+noncomputable instance :
+    LinearOrderedCancelAddCommMonoid (DegLex (α →₀ ℕ)) where
+  le_total := instLinearOrderDegLexNat.le_total
+  decidableLE := instLinearOrderDegLexNat.decidableLE
+  compare_eq_compareOfLessAndEq := instLinearOrderDegLexNat.compare_eq_compareOfLessAndEq
+
+theorem single_strictAnti : StrictAnti (fun (a : α) ↦ toDegLex (single a 1)) := by
+  intro _ _ h
+  simp only [lt_iff, ofDegLex_toDegLex, degree_single, lt_self_iff_false, Lex.single_lt_iff, h,
+    and_self, or_true]
+
+theorem single_antitone : Antitone (fun (a : α) ↦ toDegLex (single a 1)) :=
+  single_strictAnti.antitone
+
+theorem single_lt_iff {a b : α} :
+    toDegLex (Finsupp.single b 1) < toDegLex (Finsupp.single a 1) ↔ a < b :=
+  single_strictAnti.lt_iff_lt
+
+theorem single_le_iff {a b : α} :
+    toDegLex (Finsupp.single b 1) ≤ toDegLex (Finsupp.single a 1) ↔ a ≤ b :=
+  single_strictAnti.le_iff_le
+
+theorem monotone_degree :
+    Monotone (fun (x : DegLex (α →₀ ℕ)) ↦ (ofDegLex x).degree) := by
+  intro x y
+  rw [le_iff]
+  rintro (h | h)
+  · apply le_of_lt h
+  · apply le_of_eq h.1
+
+instance orderBot : OrderBot (DegLex (α →₀ ℕ)) where
+  bot := toDegLex (0 : α →₀ ℕ)
+  bot_le x := by
+    simp only [le_iff, ofDegLex_toDegLex, toLex_zero, degree_zero]
+    rcases eq_zero_or_pos (ofDegLex x).degree with (h | h)
+    · simp only [h, lt_self_iff_false, true_and, false_or, ge_iff_le]
+      exact bot_le
+    · simp [h]
+
+instance wellFoundedLT [WellFoundedGT α] :
+    WellFoundedLT (DegLex (α →₀ ℕ)) :=
+  ⟨wellFounded wellFounded_gt wellFounded_lt fun n ↦ (zero_le n).not_lt⟩
+
+section Examples
+
+/- These examples would not fit as actual `MonomialOrder` examples below
+because they use `ℕ` which doesn't have `WellFoundedGT`.
+This suggests to removing the `MonomialOrder.wf` field from the definition
+of a monomial order and add it in the hypotheses when needed. -/
```

Comment:
@joelriou what do you think of this?

Reply:
Would not these examples work with `Fin 2` instead of `ℕ`?

Reply:
they would work with `Fin 3` :-)